Inspect functionality

jacob314 opened the pull request:
I'm still cleaning up InspectorPanel and FlutterView. All other files are generally ready for review.

pq commented on src/io/flutter/utils/DiagnosticsNode.java:
```diff
@@ -0,0 +1,502 @@
+   */
+  private String dartDiagnosticId;
+
+  // XXX document
```
Possibly migrate to the `TODO(user_name):` convention?

jacob314 replied:
My convention is all XXX comments are one I expect to remove before landing.

pq commented on src/io/flutter/view/FlutterView.java:
```diff
@@ -150,6 +192,13 @@ private void updateIcon() {
         toolWindow.setIcon(ExecutionUtil.getLiveIndicator(FlutterIcons.Flutter_13));
       }
+
+      System.out.println("App changed... " + (app == null));
```
Remove `System.out`?

pq replied:
Alternatively, consider using a `Logger` or putting this debugging output behind a flag?

jacob314 replied:
Removed. That was left over debugging code.

devoncarew commented:
Awesome! Some comments on the UI; nothing that would block this PR, but things we'll want to address as follow-on work. Some are likely uncontroversial, others likely would need to be discussed:

<img width="633" alt="screen shot 2017-10-17 at 12 41 53 pm" src="https://user-images.githubusercontent.com/1269969/31685968-bcf39d72-b339-11e7-8498-13ba05c80b5c.png">

- (as noted via IM) we'll want the flutter view's toolbar always visible - so the widget and render tree tab area should go under the toolbar
- we'll want users to be able to work on their app and use the inspector side-by-side; I think we'll want to optimize the UI for a taller, narrowing amount of screen real estate. this will be somewhat challenging given the depth of the trees :)
- from use cases I've seen that the inspector can help with (from a doc from you :) ), most issues are solved via the widget tree. The render tree is used less often. Do we want to give them equal weight in the UI? This will indicate to the user that they should be using the render tree for something and could be a source of confusion. Options here are to not have a render tab - just show a widget tree under the toolbar with no tabs - or find some way to toggle to render tree mode.

devoncarew commented on src/io/flutter/utils/DiagnosticLevel.java:
```diff
@@ -0,0 +1,80 @@
+ * shown matters. Typically filtering to only show diagnostics with at least
+ * level debug is appropriate.
+ */
+public enum DiagnosticLevel {
```
This is related to APIs exposed in the flutter framework? Perhaps reference the code / classes here? We may also want the javadoc to be terser to reduce our maintenance burden, esp. if these classes are well documented in the flutter framework.

jacob314 replied:
I copied and pasted the docs from the Flutter as you guessed. Added a link to the Flutter file with the API.

devoncarew replied:
kk. going forward, we may want to tersify the docs (or eliminate them?) to reduce drift and maintenence

jacob314 replied:
I don't expect these particular classes will change much in the future. If there are changes likely it will just be adding extra logging levels in methods but I could be wrong. Hopefully these pretty carefully written docs will be helpful for other people contributing to this code on the IntelliJ. If not, they should be slimmed down or removed.

devoncarew replied:
gotcha

devoncarew commented on src/io/flutter/utils/DiagnosticsNode.java:
```diff
@@ -0,0 +1,502 @@
+   * Only applies to IterableProperty.
+   */
+  public ArrayList<String> getValues() {
+    ArrayList<String> values = null;
```
probably don't need to init this

jacob314 replied:
I get a "values" might not have been initialized java lint warning otherwise.
Cleaned up the method to use an early return style to reduce nesting and avoid the = null.

devoncarew replied:
odd

devoncarew commented on src/io/flutter/utils/DiagnosticsNode.java:
```diff
@@ -0,0 +1,502 @@
+    }
+    return properties;
+  }
+}
```
just to reduce future diffs, I would do a formatting pass on the files

jacob314 replied:
done

devoncarew commented on src/io/flutter/utils/EvalOnDartLibrary.java:
```diff
@@ -0,0 +1,195 @@
+    try {
+      final Class<?> c = Class.forName("org.dartlang.vm.service.VmServiceBase");
+
+      final Method[] allMethods = c.getDeclaredMethods();
```
There is a `com.intellij.util.ReflectionUtil` class which can reduce a bit of the work here (used for example by `FlutterReloadManager`)

jacob314 replied:
That is handy. Done.

devoncarew commented on src/io/flutter/utils/EvalOnDartLibrary.java:
```diff
@@ -0,0 +1,195 @@
+           @Override
+           public void onError(RPCError error) {
+             // TODO(jacobr): what is the right way to log this case?
+             System.err.println("Got RPCError: " + error);
```
Presumably the client will listen to the error result and display (or ignore) appropriately to the user? If you do need to log, you can use the intellij logging system: https://github.com/flutter/flutter-intellij/blob/335e25cc3c97d515305ec72ec4dd48dd3ebba8e8/src/io/flutter/run/daemon/DaemonEvent.java#L290

jacob314 replied:
Done

jacob314 commented on src/io/flutter/utils/DiagnosticLevel.java:
```diff
@@ -0,0 +1,80 @@
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+package io.flutter.utils;
```
moved to a new ` .inspector.` package

devoncarew commented:
Two other thoughts:

- post-landing, we should update the release testing script (https://github.com/flutter/flutter-intellij/blob/master/docs/testing.md) to include a brief smoke-test of the inspector (brevity is important to make sure the overall script takes < 10 min)
- per a discussion, most of the automated testing of this will happen once we have general UI integration test support and a headless flutter device available. Looking over this PR, I think the json parsing code is amenable to unit tests; perhaps file an issue to add unit tests for these?

jacob314 commented:
PTAL. Added a setting menu item to configure whether the Inspector is displayed. Moved the inspector so it is the main panel. Made the toolbar show on both panels.

stevemessick commented on src/io/flutter/inspector/DiagnosticLevel.java:
```diff
@@ -0,0 +1,83 @@
+   * If a user chooses to display [hidden] diagnostics, they should not expect
+   * the diagnostics to be formatted consistently with other diagnostics and
+   * they should expect them to sometimes be be misleading. For example,
+   * [FlagProperty] and [ObjectFlagProperty] have uglier formatting when the
```
Did we adopt a new javadoc convention? I'm not familiar with [...] in javadoc. I thought we used {@code ...}. If those references should be rewritten then I think it can wait until a later PR.

jacob314 replied:
These are copy and pasted from the Dart comments for these properties so the style is a bit wrong. I'll clean it up with a followup PR.

stevemessick commented on src/io/flutter/inspector/EvalOnDartLibrary.java:
```diff
@@ -0,0 +1,185 @@
+                       new EvaluateConsumer() {
+                         @Override
+                         public void onError(RPCError error) {
+                           // TODO(jacobr): what is the right way to log this case?
```
Are you wondering if an error here is serious enough to bring to the user's attention? If so, then maybe an error balloon in the style of IntelliJ would work.

jacob314 replied:
Removed the TODO as the LOG.error call is sufficient for this unlikely exceptional case.

devoncarew commented on resources/META-INF/plugin.xml:
```diff
@@ -299,6 +299,18 @@
+      </action>
+      <action id="Flutter.JumpToSource" class="io.flutter.inspector.JumpToSourceAction"
+              text="Jump to Source">
+        <keyboard-shortcut keymap="$default" first-keystroke="control DOWN"/>
```
If this is not frequently used, we may not want a keybinding.

devoncarew replied:
If this doesn't conflict (as per the item above), then no real need to remove.

jacob314 replied:
I think we might as well be consistent with the existing debugger panel keyboard shortcuts that do the same thing. Personally I find having keybindings for these handy as jumping to the source is really handy.

devoncarew replied:
makes sense

pq commented on src/io/flutter/inspector/InspectorService.java:
```diff
@@ -0,0 +1,352 @@
+   * Converts an inspector ref to value suitable for use by generic intellij
+   * debugging tools.
+   * <p>
+   * Warning: DartVmServiceValue references do not make any lifetime gaurantees
```
gaurantees => guarantees

jacob314 replied:
done

jacob314 commented on resources/META-INF/plugin.xml:
```diff
@@ -299,6 +299,18 @@
+      <action id="Flutter.JumpToTypeSource" class="io.flutter.inspector.JumpToTypeSourceAction"
+              description="Jump to Type Source"
+              text="Jump to Type Source">
+        <keyboard-shortcut keymap="$default" first-keystroke="shift F4"/>
```
I've verified it doesn't conflict with the existing binding. I believe the reason is that the action is only activated within the context of the inspector panel not across the whole application.

devoncarew replied:
👍

jacob314 commented on src/io/flutter/inspector/DiagnosticsNode.java:
```diff
@@ -0,0 +1,511 @@
+    return getStringMember("type");
+  }
+
+  // The follow fields are only specified for properties.
```
Removed this comment and made sure all fields relevant to properties have comments specifying what kind of properties they apply to.

brianegan commented:
Whaaaaaat! This is getting next level :D As a user, this looks amazing. Nice work!

pq commented:
🎉 🎂 🎉
